File: elixir/bob/bob.exs
defmodule Bob do
  def hey(input) do
    input = String.trim(input)
    cond do
      String.length(input) == 0 ->
        "Fine. Be that way!"
      yelling?(input) && questioning?(input) ->
        "Calm down, I know what I'm doing!"
      yelling?(input) ->
        "Whoa, chill out!"
      questioning?(input) ->
        "Sure."
      true -> "Whatever."
    end
  end

  defp questioning?(input) do
    String.ends_with?(input, "?")
  end

  defp yelling?(input) do
    includes_letter?(input) && String.upcase(input) == input
  end

  defp includes_letter?(input) do
    String.upcase(input) != String.downcase(input)
  end
end

File: ruby/bob/bob.rb
class Bob
  def self.hey(sentence)
    new(sentence).answer
  end

  def answer
    if silence?
      'Fine. Be that way!'
    elsif question? and yell?
      "Calm down, I know what I'm doing!"
    elsif question?
      'Sure.'
    elsif yell?
      'Whoa, chill out!'
    else
      'Whatever.'
    end
  end

  private

  attr_reader :sentence

  def initialize(sentence)
    @sentence = sentence.strip
  end

  def question?
    sentence.end_with? '?'
  end

  def yell?
    sentence.upcase == sentence and letters?
  end

  def silence?
    sentence.empty?
  end

  def letters?
    sentence.downcase != sentence.upcase
  end
end
